File: web/list/list.go
package list

import (
	"ily.dev/act3/html"
	"ily.dev/act3/html/attr"
	. "ily.dev/act3/ui"
	"ily.dev/act3/ui/stimulus"
)

const (
	controller = "list"
)

var (
	ID  = attr.Attr("data-list-id-param")
	URL = attr.Attr("data-list-url-param")
)

func List(prefix, target string, attrs ...attr.Node) html.Element {
	return ScrollArea(
		attr.Class("p-2"),
		attr.Group(attrs...),
		stimulus.Controller(controller),
		stimulus.Value(controller, "prefix")(prefix),
		stimulus.Value(controller, "target")(target),
		stimulus.Action("turbo:render@document->list#render"),
	)
}

func Items[T any](items []T, f func(T, ...attr.Node) html.Node) html.Node {
	return html.Range(items, func(v T) html.Node {
		return f(v,
			stimulus.Target("list", "item"),
			stimulus.Action("click->list#select"),
		)
	})
}

File: ui/list.js
import { Controller } from "../web/stimulus.js";

export default class extends Controller {
	static targets = ["item"];
	static values = {
		prefix: String,
		target: String,
		selected: String,
	};

	initialize() {
		this.#initSelected();
		this.#showSelected();
	}

	selectedValueChanged() {
		this.#showSelected();
	}

	select(ev) {
		this.selectedValue = ev.params.id;
		this.#navigate();
	}

	render() {
		this.#initSelected();
	}

	#initSelected() {
		const path = document.location.pathname;
		if (!path.startsWith(this.prefixValue)) {
			this.selectedValue = "";
			return;
		}
		for (const t of this.itemTargets) {
			const url = t.getAttribute("data-list-url-param");
			if (path === url || path.startsWith(url + "/")) {
				this.selectedValue = t.getAttribute("data-list-id-param");
				return;
			}
		}
		this.selectedValue = "";
	}

	#navigate() {
		if (this.selectedValue === "") return;
		for (const t of this.itemTargets) {
			if (t.getAttribute("data-list-id-param") === this.selectedValue) {
				const url = t.getAttribute("data-list-url-param");
				Turbo.visit(url, { frame: this.targetValue });
				return;
			}
		}
	}

	#showSelected() {
		for (const t of this.itemTargets) {
			const id = t.getAttribute("data-list-id-param");
			if (id === this.selectedValue) {
				t.setAttribute("data-selected", "");
			} else {
				t.removeAttribute("data-selected");
			}
		}
	}
}
